File: cscs-checks/prgenv/src/mpi_openmp/hello_world_mpi_openmp.f90
program hello_world_mpi_openmp
use omp_lib
implicit none

include 'mpif.h'

integer :: rank, size, ierr, i, j, k, id, nthreads
integer, dimension(MPI_STATUS_SIZE) :: status

integer, allocatable :: buffer(:), output(:)

call MPI_INIT(ierr)
call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)
call MPI_COMM_SIZE(MPI_COMM_WORLD, size, ierr)

!$OMP PARALLEL SHARED(nthreads)
    !omp critical
      nthreads = omp_get_num_threads()
    !omp end critical
!$OMP END PARALLEL

allocate(buffer(nthreads))
allocate(output(nthreads*size))

! Populating the buffer with -1, this will be used for checking
buffer = -1
output = -1

! Populating the buffer
!$OMP PARALLEL PRIVATE(id, nthreads) SHARED(BUFFER, OUTPUT)
      id = omp_get_thread_num()
      buffer(id+1) = id
!$OMP END PARALLEL

CALL MPI_Gather(buffer, nthreads, MPI_INTEGER, &
                output, nthreads, MPI_INTEGER, &
                0, &
                MPI_COMM_WORLD, ierr)

if (rank .eq. 0) then
    do i = 1, size
       do j = 1, nthreads
         k = (nthreads * (i - 1)) +j
         write (*,'(A23,1X,I3,1X,A6,1X,I3,1X,A12,1X,I3,1X,A6,I3)') 'Hello World from thread', output(k), &
                  'out of', nthreads, 'from process', i-1, 'out of', size
       enddo
    enddo
endif

deallocate(buffer)
deallocate(output)

call MPI_FINALIZE(ierr)

end program hello_world_mpi_openmp

File: cscs-checks/prgenv/src/mpi_openmp/hello_world_mpi_openmp.cpp
#include <stdio.h>
#include <omp.h>
#include "mpi.h"

int main(int argc, char *argv[])
{
  int size, rank;
  // int namelen;
  // char processor_name[MPI_MAX_PROCESSOR_NAME];
  int tid = 0;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &size);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  // MPI_Get_processor_name(processor_name, &namelen);

  #pragma omp parallel default(shared) private(tid)
  {
    int nthreads = omp_get_num_threads();
    tid = omp_get_thread_num();
    printf("Hello, World from thread %d out of %d from process %d out of %d\n",
           tid, nthreads, rank, size);
  }

  MPI_Finalize();

  return 0;
}
